src/garage/sampler/base.py

```diff
@@ -1,43 +1,48 @@
+"""Base sampler class."""
+
 import abc
+import copy
+
+from garage.sampler.config import SamplerConfig
 
 
 class Sampler(abc.ABC):
-    @abc.abstractmethod
+    """Base class of all samplers."""
+
+    def __init__(self, algo, env):
+        """Construct a Sampler from an Algorithm.
+
+        Using this method is deprecated.
+        """
+        self.algo = algo
+        self.env = env
+
+    @classmethod
+    def construct(cls, config: SamplerConfig, agent, env) -> 'Sampler':
```

Comment:
if we have this construct that takes a config, can we get rid of passing the algo to the init @krzentner

Reply:
NVM

Reply:
Yeah, the goal of this change is actually to let us avoid the `Sampler`s ever having direct access to `Algorithm`s. They only get `SamplerConfig`s.

Reply:
> ... if the init is deprecated, are you saying that construct is going to be the new init of the sampler?

The intent is actually to have algorithms ask the runner to create samplers. This is because the runner generally knows the correct / default values for `init_worker_fn` and `update_agent_fn`. But yeah, the runner will construct Samplers by calling `construct`, at least for now. In the future, we *could* refactor things so that `__init__` is used by the runner instead, but we might not want to do that, since there's a general aversion to making `__init__` part of APIs.